### ISSUE TITLE
Gemma-4 PLE tensors silently dropped on `--quant none` extracts → garbage INFER

### ISSUE THREAD

**Issue by codeninja** [chrishayuk/larql#49]
# Gemma-4 with PLE active produces garbage INFER on `--quant none` (f16/f32) extracts

## TL;DR

`larql extract` against Gemma-4 models that activate Per-Layer Embeddings
(`hidden_size_per_layer_input > 0`, e.g. `gemma-4-E4B-it`) silently produces
a vindex with no PLE tensors when `--quant none`/`--quant f16` is used. Inference
then runs without the PLE additive contribution, producing near-random outputs
(top-5 garbage tokens at ~5-9% confidence each). The Q4_K writer already
handles this correctly — the bug is specifically in `write_f32.rs`.

## Reproducer

```bash
# Background: gemma-4-E4B-it has hidden_size_per_layer_input=256 → PLE active
larql extract google/gemma-4-E4B-it -o e4b.vindex                 # default --quant none
larql lql 'USE "e4b.vindex"; INFER "The capital of France is" TOP 5;'
```

Expected: `Paris` at high confidence.
Observed:
```
Predictions (walk FFN):
   1.  mutable             (8.92%)
   2. ッケ                   (7.39%)
   3. ceral                (7.17%)
   4.  ఉ                   (4.77%)
   5. sby                  (4.76%)
```

The same model extracted with `--quant q4k` works correctly (the Q4_K writer
already captures PLE tensors into `ple_weights.bin`).

### 1. Original safetensors contain 129 PLE tensors

```python
$ python3 -c "from safetensors import safe_open; f=safe_open('.../model.safetensors','pt'); 
              keys=list(f.keys()); ple=[k for k in keys if 'per_layer' in k.lower() or 'embed_tokens_per' in k.lower()];
              print(f'total: {len(keys)}, PLE: {len(ple)}')"
total: 2130, PLE: 129
```

The 129 are 3 globals (`embed_tokens_per_layer`, `per_layer_model_projection`,
`per_layer_projection_norm`) plus per-layer triplets (`per_layer_input_gate`,
`per_layer_projection`, `post_per_layer_input_norm`) × 42 layers.

### 2. The `--quant none` vindex's `weight_manifest.json` has zero PLE entries

```python
$ python3 -c "import json; m=json.load(open('e4b.vindex/weight_manifest.json'));
              ple=[e for e in m if 'per_layer' in e['key'].lower() or 'embed_tokens_per' in e['key'].lower() or 'layer_scalar' in e['key']];
              print(f'total: {len(m)}, PLE+layer_scalar: {len(ple)}')"
total: 505, PLE+layer_scalar: 0
```

Also note: zero `layer_scalar` entries, even though Gemma-4 forward applies
`layer_scalar` per layer when present.

### 3. `crates/larql-vindex/src/format/weights/write_q4k/mod.rs` HAS the PLE path

Lines ~474-563 explicitly:
- Write `ple_weights.bin` with `embed_tokens_per_layer.weight`,
  `per_layer_model_projection.weight`, per-layer `per_layer_input_gate.weight`,
  `per_layer_projection.weight` as `kind::TENSOR_F16`.
- Add `per_layer_projection_norm.weight`, `post_per_layer_input_norm.weight`,
  `layer_scalar` to `norms.bin`.

### 4. `crates/larql-vindex/src/format/weights/write_f32.rs` has zero such handling

A grep for `per_layer|ple_weights|embed_tokens_per|layer_scalar` returns one
match: line 589 referencing `per_layer_embed_dim` for the `model_config`
metadata only.

### 5. The forward-pass guards silently no-op on missing tensors

[`crates/larql-inference/src/forward/ple.rs:37-39`](crates/larql-inference/src/forward/ple.rs)
in `precompute_per_layer_inputs`:

```rust
let w_model_proj = match weights.tensors.get("per_layer_model_projection.weight") {
    Some(w) => w,
    None => return Vec::new(),    // silent skip
};
```

[`crates/larql-inference/src/forward/ple.rs:117`](crates/larql-inference/src/forward/ple.rs)
in `apply_per_layer_embedding`:

```rust
let per_layer_input = match per_layer_input {
    Some(p) => p,
    None => return h.clone(),    // silent skip
};
```

So when PLE tensors aren't loaded, the forward pass falls through with no
warning and no error. Garbage output is the only signal.

### 6. The CPU goldens never exercised PLE

[`crates/larql-inference/tests/test_logits_goldens.rs`](crates/larql-inference/tests/test_logits_goldens.rs)
pins `gemma4-31B-it (dense)` CPU top-1 logit at `2.366634`. But 31B's config
has `hidden_size_per_layer_input: 0` — PLE is disabled — so the codepath was
never validated by the test suite. No golden for `gemma-4-E4B-it` (PLE active)
or `gemma-4-26B-A4B-it` (MoE, also affected by a separate WalkFfn gap).

## Why this isn't CPU vs Metal

The dropped-tensor data gap will silently break Metal inference of E4B too —
Metal reads the same vindex written by the same writer. The bug is in the
extract+load pipeline, upstream of any backend selection.

### 1. `crates/larql-vindex/src/format/weights/write_f32.rs`

Add Gemma-4 sidecar writing. Mirror the existing Q4_K behavior — write 2-D
PLE tensors to `ple_weights.bin` as `kind::TENSOR_F16`, with manifest entries
for:

- `per_layer_model_projection.weight`
- `embed_tokens_per_layer.weight`
- `layers.N.per_layer_input_gate.weight`
- `layers.N.per_layer_projection.weight`

Plus add to `norms.bin`:

- `layers.N.layer_scalar`
- `layers.N.post_per_layer_input_norm.weight`
- `per_layer_projection_norm.weight`

(`post_per_layer_input_norm` belongs with norms, not with the larger PLE
tensors — same layout choice as the Q4_K writer.)

### 2. Refactor: shared helper for PLE/sidecar writing

Move the PLE-writing block from `write_q4k/mod.rs` into a shared helper in
`crates/larql-vindex/src/format/weights/` (e.g. `ple_sidecar.rs`) used by
both writers. Avoids future drift.

### 3. `crates/larql-vindex/src/format/weights/load.rs`

Teach the full float loader to load `kind::TENSOR_F16` into
`weights.tensors`, matching the Q4 loader. Then add post-load validation:

```rust
if arch.has_per_layer_embeddings() {
    require!("per_layer_model_projection.weight", weights.tensors);
    require!("embed_tokens_per_layer.weight", weights.tensors);
    for layer in 0..num_layers {
        require!(format!("layers.{layer}.per_layer_input_gate.weight"), …);
        // …per_layer_projection, post_per_layer_input_norm…
    }
    if arch.is_gemma4() {
        for layer in 0..num_layers {
            require!(format!("layers.{layer}.layer_scalar"), weights.vectors);
        }
    }
}
```

Return `VindexError::Parse` with a "rebuild this vindex with current
larql — old extract dropped Gemma-4 sidecars" message.

### 4. `crates/larql-inference/src/forward/ple.rs`

Replace the silent `return Vec::new()` / `return h.clone()` paths with
`expect()` messages. Once the loader validates the invariant, missing
tensors here is a panic-worthy bug, not a fall-through case.

The "no PLE configured" branch (line 27-29's `arch.has_per_layer_embeddings()`
check) stays — that's a real architectural state.

### 5. `crates/larql-vindex/tests/test_vindex.rs`

Synthetic non-Q4 streaming extraction test against a tiny Gemma-4 PLE config.
Asserts:

- `ple_weights.bin` exists in the output dir
- Manifest contains the expected PLE `tensor_f16` entries
- `norms.bin` contains the PLE norms + `layer_scalar` entries
- `load_model_weights` populates `weights.tensors` / `weights.vectors`

Update the existing Q4_K PLE test to also assert `layer_scalar` presence
(currently uncovered there too).

### 6. `crates/larql-inference/tests/test_logits_goldens.rs`

Add an `#[ignore]` real-vindex CPU golden for `gemma-4-E4B-it` after a fresh
extract. The current harness assumes Q4_K everywhere — branch on
`cfg.quant` so a non-Q4 vindex flows through `InferenceWeights::infer_patched`
correctly. Pin token IDs and top-1 logit (not decoded strings — the same
seed reproduces with bit-equal float behavior under the same backend).

## Adjacent issues surfaced during review

These are **separate bugs** not covered by this fix; logging here so they
don't get lost:

### A. WalkFfn doesn't implement `forward_moe_full_layer`

`crates/larql-inference/src/vindex/walk_ffn/mod.rs` has zero MoE references.
On a hybrid-MoE Gemma-4 (`gemma-4-26B-A4B-it`), the dispatch in
`predict_q4k_hidden_with_ffn`:

```rust
if weights.arch.is_hybrid_moe() {
    if let Some(h_post_attn) = run_attention_public(weights, &h, layer) {
        if let Some(h_out) = ffn_backend.forward_moe_full_layer(layer, &h_post_attn) {
            h = h_out; continue;
        }
    }
}
// fall through to dense run_layer_with_ffn
```

`WalkFfn::forward_moe_full_layer` returns `None` (default trait impl), so
MoE Gemma-4 silently falls back to the dense walk — wrong outputs.

**Fix scope:** implement `forward_moe_full_layer` on `WalkFfn` using the
existing `cpu_moe_forward` (already used by `predict_q4k_hidden`). Or, more
narrowly: route LQL INFER through `predict_q4k_hidden` directly when the
model is hybrid-MoE.

### B. `predict_with_ffn_attention` doesn't thread shared KV

Suspect `EXPLAIN INFER WITH ATTENTION` is incorrect on E4B after layers that
share KV. Out of scope here; mention in the PR for follow-up.

### C. LQL INFER tokenization doesn't chat-wrap

`exec_infer` calls `tokenizer.encode(prompt, true)` directly. The golden
harness uses `larql_inference::wrap_chat_prompt`. For -it (instruct) models,
this can produce different first-token distributions even when the model is
otherwise correct. Worth adding chat wrapping to the LQL path so a manual
`larql lql 'INFER "..."'` matches what the goldens validate.

## Why it slipped past review

- Goldens pinned only on PLE-disabled Gemma-4 (31B).
- Both the writer guard (`arch.has_per_layer_embeddings()`) and the loader
  guard (`weights.tensors.get(...)? else None`) silently no-op.
- Default extraction format on the tooling docs is implicit `--quant none`
  (the README's first example uses no `--quant` flag), so users naturally
  hit the broken path.

## Mitigation while the fix lands

`larql extract <gemma-4-with-PLE-model> --quant q4k` works today. The Q4_K
writer captures PLE; the Q4 loader hydrates it. The visualizer (or any
INFER-using consumer) can be pointed at a q4k vindex and gets correct
predictions immediately.

---

## Credits

Diagnosis by Claude Opus 4.7 (initial investigation, narrowed from "MoE bug"
hypothesis to "PLE silent drop"). Refined by Codex CLI 0.125.0 (caught the
write_f32 vs write_q4k asymmetry, plus `layer_scalar`, plus the LQL chat-wrap
hypothesis). Phase 0 mitigation (re-extract with `--quant q4k`) verified
against `google/gemma-4-E4B-it`.

**Comment by deem0n** [comment]
I will try to fix it.

**Comment by deem0n** [comment]
@codeninja can you test the fix in #91?
